### PR TITLE
Support explicitly setting a dimension order with to_dataframe()

### LINKED ISSUE #4331: Support explicitly setting a dimension order with to_dataframe()
As discussed in https://github.com/pydata/xarray/issues/2346, it would be nice to support explicitly setting the desired order of dimensions when calling `Dataset.to_dataframe()` or `DataArray.to_dataframe()`.

There is nice precedent for this in the `to_dask_dataframe` method:
http://xarray.pydata.org/en/stable/generated/xarray.Dataset.to_dask_dataframe.html

I imagine we could copy the exact same API for `to_dataframe.

### PR DESCRIPTION
**Pull request by Thomas-Z**
- [x] Closes #4331
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by max-sixty** [comment]
Great, thanks @Thomas-Z , LGTM

Does anyone have thoughts on the kwarg name? `dim_order` seems reasonable, though IIRC we don't use it elsewhere

**Comment by dcherian** [comment]
How about `dims: Iterable[Hashable]`?

**Comment by Thomas-Z** [comment]
Hello,

I actually followed @shoyer [suggestion](https://github.com/pydata/xarray/issues/4331) to use [to_dask_dataframe](http://xarray.pydata.org/en/stable/generated/xarray.Dataset.to_dask_dataframe.html) parameter name.

**Comment by dcherian** [comment]
OK consistency with `to_dask_dataframe` is a good idea. Maybe stick with `dim_order` for now and we can change both later if someone really wants it?

**Comment by shoyer** [comment]
I like `dim_order` a little bit better than `dims` because it's clear that you need supply all the dimensions, not just a subset (which is usually the case for `dims`)

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -4533,14 +4533,37 @@ def _to_dataframe(self, ordered_dims):
...
         """
-        return self._to_dataframe(self.dims)
+
+        if dim_order is None:
+            dim_order = list(self.dims)
+        elif set(dim_order) != set(self.dims):
+            raise ValueError(
+                "dim_order {} does not match the set of dimensions on this "
+                "Dataset: {}".format(dim_order, list(self.dims))
+            )
```
Comment (shoyer) [inline xarray/core/dataset.py:4564]: Maybe we could make a small helper method `_normalize_dim_order()` that we could use both here and in `to_dask_dataframe`?

**Comment by Thomas-Z** [comment]
Do we want DataArray.to_dataframe to be consistent with Dataset.to_dataframe regarding the default dimension ordering (i.e. alphabetically) or do we want to keep the current behavior (DataArray.dims order)?

**Comment by shoyer** [comment]
> Do we want DataArray.to_dataframe to be consistent with Dataset.to_dataframe regarding the default dimension ordering (i.e. alphabetically) or do we want to keep the current behavior (DataArray.dims order)?

`DataArray.to_dataframe()` should keep the current behavior based on the dimension order.

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -2381,15 +2404,20 @@ def to_dataframe(self, name: Hashable = None) -> pd.DataFrame:
...
         # By using a unique name, we can convert a DataArray into a DataFrame
         # even if it shares a name with one of its coordinates.
         # I would normally use unique_name = object() but that results in a
         # dataframe with columns in the wrong order, for reasons I have not
         # been able to debug (possibly a pandas bug?).
         unique_name = "__unique_name_identifier_z98xfz98xugfg73ho__"
         ds = self._to_dataset_whole(name=unique_name)
-        df = ds._to_dataframe(dims)
+
+        if dim_order is None:
```
Comment (dcherian) [inline xarray/core/dataarray.py:2415]: can we handle the `None` case in `_normalize_dim_order`?

Comment (shoyer) [reply xarray/core/dataarray.py:2415]: We need it here because `DataArray.to_dataframe()` uses a different dimension order than `Dataset.to_dataframe()`

Comment (dcherian) [reply xarray/core/dataarray.py:2415]: oh thanks!

**Comment by dcherian** [comment]
Thanks @Thomas-Z . I see this is your first PR here. Welcome to xarray!

**Comment by max-sixty** [comment]
Thanks @Thomas-Z ! Great to have you as a contributor

**Comment by Thomas-Z** [comment]
My pleasure.
I've been a user for a few years now, I'll gladly give something back whenever I can.

**Comment by hammer** [comment]
Just noting for GitHub metadata purposes that this PR addresses the `Dataset.to_dataframe` annotation request in https://github.com/pydata/xarray/issues/4238

**Pull request merged**